From 078700b65f4d5a5e3a699a774b13d4461163e6e0 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:06:53 -0400
Subject: [tools] perf/hists: Add missing update on filtered stats in hists__decay_entries()

Message-id: <1410181752-25631-169-git-send-email-jolsa@redhat.com>
Patchwork-id: 91850
O-Subject: [PATCH RHEL7.1 BZ1134356 168/307] perf hists: Add missing update on filtered stats in hists__decay_entries()
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 3186b6815d49b5e0defbd884223da3778edb59fc
Author: Namhyung Kim <namhyung@kernel.org>
Date: Tue Apr 22 13:44:23 2014 +0900

description
===========
When a filter is used for perf top, its hists->nr_non_filtered_entries
was not updated after it removed an entry in hists__decay_entries().
Also hists->stats.total_non_filtered_period was missed too.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/hist.c b/tools/perf/util/hist.c
index 6d0d2d7..7f0236c 100644
--- a/tools/perf/util/hist.c
+++ b/tools/perf/util/hist.c
@@ -225,14 +225,18 @@ static void he_stat__decay(struct he_stat *he_stat)
 static bool hists__decay_entry(struct hists *hists, struct hist_entry *he)
 {
  u64 prev_period = he->stat.period;
+ u64 diff;
 
  if (prev_period == 0)
   return true;
 
  he_stat__decay(&he->stat);
 
+ diff = prev_period - he->stat.period;
+
+ hists->stats.total_period -= diff;
  if (!he->filtered)
-  hists->stats.total_period -= prev_period - he->stat.period;
+  hists->stats.total_non_filtered_period -= diff;
 
  return he->stat.period == 0;
 }
@@ -259,8 +263,11 @@ void hists__decay_entries(struct hists *hists, bool zap_user, bool zap_kernel)
    if (sort__need_collapse)
     rb_erase(&n->rb_node_in, &hists->entries_collapsed);
 
-   hist_entry__free(n);
    --hists->nr_entries;
+   if (!n->filtered)
+    --hists->nr_non_filtered_entries;
+
+   hist_entry__free(n);
   }
  }
 }
-- 
1.7.1